### PR TITLE
Prevent fdbcli status drift

### PR DESCRIPTION
**Pull request by manfontan**
# Description

Update strategy in some cases may be to aggressive deleting pods. This is because the Fault Tolerance check is not very restrictive. Checking only the status json for availability. Adding additional safeguards like checking the POD state would make this verification more reliable. Preventing the operator from replacing too many pods while others are still not running.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Discussion

The idea here is to compare the status json output with the cluster status. To make sure they have the same processes. If required we could make a more fined grained comparison but this would be a first step. If any of the processes returned by the fdbcli status call don't have a matching pod we will re-queue and to try again later.

## Testing

*Please describe the tests that you ran to verify your changes. Unit tests?
Manual testing?*

*Do we need to perform additional testing once this is merged, or perform in a larger testing environment?*

## Documentation

*Did you update relevant documentation within this repository?*

*If this change is adding new functionality, do we need to describe it in our user manual?*

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*

*Does this introduce new defaults that we should re-evaluate in the future?*

### CODE REVIEW

**Review by johscheuer — COMMENTED** [review]
I know this is still a draft but I wanted to comment on what I think makes sense to check. We also check the fault tolerance in the [HasDesiredFaultTolerance](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/internal/fault_tolerance.go) method (in addition to the availability).

File: controllers/update_pods.go [hunk 1]
```diff
@@ -164,6 +173,22 @@ func shouldRequeueDueToTerminatingPod(pod *corev1.Pod, cluster *fdbv1beta2.Found
...
 
+// Because the internal.fault_tolerance relies in fdbcli status only we need to make sure it is
+// synced with the cluster status to prevent race conditions.
+func shouldRequeueDueToClusterStatusDrift(cluster *fdbv1beta2.FoundationDBCluster, adminClient fdbadminclient.AdminClient, podMap map[string]*corev1.Pod) (bool, error) {
+	status, err := adminClient.GetStatus()
+	if err != nil {
+		return false, err
+	}
+	//Check that all processes in fdbcli status have a matching pod.
+	for _, p := range status.Cluster.Processes {
```
[the remark this answers is not in the document]
Comment (johscheuer) [reply controllers/update_pods.go:184]: Thinking more about this case it might be enough to just check if any Pod has a non zero `deletionTimestamp` and is not marked for removal. That would indicate that a deletion/update is already in progress and we want to skip further steps.

[the remark this answers is not in the document]
Comment (manfontan) [reply controllers/update_pods.go:184]: So we would be adding a check like `shouldRequeueDueToTerminatingPod` but instead of checking `cluster.ProcessGroupIsBeingRemoved` we need to check if the process group is not marked for removal right?

File: controllers/update_pods.go [hunk 2]
```diff
@@ -164,6 +173,22 @@ func shouldRequeueDueToTerminatingPod(pod *corev1.Pod, cluster *fdbv1beta2.Found
...
+		if _, ok := podMap[p.Locality["instance_id"]]; !ok {
```
Comment (johscheuer) [inline controllers/update_pods.go:185]:
```suggestion
		if _, ok := podMap[p.Locality[fdbv1beta2.FDBLocalityInstanceIDKey]; !ok {
```

See: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdb_labels.go#L62

File: controllers/update_pods.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (johscheuer) [reply controllers/update_pods.go:184]: Looking again at `shouldRequeueDueToTerminatingPod` that should actually cover the case that you've describe and the default time we block any new updates is 10min. That change was added a while back: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1070 and should be in all versions of 1.0.0+. One thing that we are not checking is if the new Pods are up and running before moving to the next Pod: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/update_pods.go#L71-L76 but that should be catched in the fault tolerance check (since I would expect that FT is reduced in that case).

**Comment by manfontan** [comment]
Hi @johscheuer 

Please accept my apology for the delay in my response and thank you for all your insights. They are really helpful. 

> If you could provide some logs or additional information that shows that is a valid case, we might be able to help work on a potential solution.

I have reviewed the logs for the `fdb-kubernetes-operator-controller-manager` covering the time of the incident but I cannot find any obvious issues. We are currently running a 1.4.x version of the operator so fault tolerant check logging is not reported. I plan to upgrade to at least v1.5.0 to get fault tolerant check logs. Hopefully this will help. 

I cannot provide them on github unfortunately. If there is a safe way of sharing the logs that could be an option, but I should likely require approval to share them.

> Just out of curiosity was there and incident/issue where you observed this behaviour and are you able to share some information. about it? just to better understand the motivation behind this PR.

Yes. We have seen our number of replicas going down to 1 several times. Our expectation is that it should never go below 2. Since we have triple replication in place. When the update process starts we expect to the fault tolerance to be reduced by 1. Then the Operator should block additional deletes until the fault tolerance is restored. In the latest incident we saw the number of replicas fall down to 0.

In summary we will upgrade to v1.5 or later and I will monitor the next upgrades to gather logs (I am currently looking at the controller manager logs) and metrics in order to get a better picture of what is causing this issue. If you have any suggestions or additional information is required to troubleshoot this issue please let me know.

**Comment by simenl** [comment]
Hi! I reproduced this issue with version 1.8.1 of the fdb-kubernetes-operator, see the operator logs in the drop down.

**Setup** 
```
Spec:
  Database Configuration:
    Logs:                             5
    Proxies:                          3
    redundancy_mode:                  triple
    Storage:                          5
    storage_engine:                   ssd
  Minimum Uptime Seconds For Bounce:  600
  Process Counts:
    Log:        5
    Stateless:  3
    Storage:    5

Running Version:  6.3.12
```

    
To make sure that no recreated pods could be scheduled, I cordoned all the nodes. `kubectl cordon <node>`. I then triggered a rolling bounce pod update strategy (by changing the sidecar image tag).

We can see the reported fault tolerance from `status/json` is 2 just after two of the storage pods are detected as MissingProcesses (one of them also reporting PodPending). The kubernetes operator ends up deleting and recreating three pods (as pending), before the reported fault tolerance from `status/json` goes below 2.

End result is the cluster being unavailable, as a whole team of pods are not running (3 out of 5 storage pods are pending). 

<details><summary>Operator Logs</summary>

```json
{"level":"info","ts":1665001774.3598943,"logger":"controller","msg":"Attempting to run sub-reconciler","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.updateStatus"}
{"level":"info","ts":1665001774.360013,"logger":"fdbclient","msg":"Fetch values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001774.4088979,"logger":"fdbclient","msg":"Done fetching values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001774.4784985,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001774.4785688,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001774.47859,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001774.4786031,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001774.4786162,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
...
{"level":"info","ts":1665001774.8717215,"logger":"controller","msg":"Deleting pods","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","zone":"26081535-vmss0000yt","count":1,"deletionMode":"Zone"}
{"level":"info","ts":1665001774.9156864,"logger":"controller","msg":"Delaying requeue for sub-reconciler","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.updatePods","message":"Pods need to be recreated","error":null}
{"level":"info","ts":1665001774.9158194,"logger":"controller","msg":"Attempting to run sub-reconciler","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.removeProcessGroups"}
{"level":"info","ts":1665001774.9160814,"logger":"controller","msg":"Attempting to run sub-reconciler","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.removeServices"}
{"level":"info","ts":1665001774.9162133,"logger":"controller","msg":"Attempting to run sub-reconciler","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.updateStatus"}
...
{"level":"info","ts":1665001774.9581664,"logger":"fdbclient","msg":"Done fetching values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001777.0301375,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001777.0301926,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001777.0302026,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001777.030211,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001777.0302234,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","SidecarUnreachable"]}
...
{"level":"info","ts":1665001809.3765922,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001809.3766372,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001809.3766472,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001809.3766623,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001809.3766763,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","SidecarUnreachable","MissingPod"]}
...
{"level":"info","ts":1665001831.3356743,"logger":"fdbclient","msg":"Fetch values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001835.3488076,"logger":"fdbclient","msg":"Done fetching values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001835.3498478,"logger":"controller","msg":"Check desired fault tolerance","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","expectedFaultTolerance":2,"maxZoneFailuresWithoutLosingData":2,"maxZoneFailuresWithoutLosingAvailability":2}
{"level":"info","ts":1665001835.3498828,"logger":"controller","msg":"Taking lock on cluster","namespace":"namespace","cluster":"foundationdb-cluster","action":"updating pods"}
{"level":"info","ts":1665001835.3498986,"logger":"controller","msg":"Deleting pods","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","zone":"26081535-vmss0000y4","count":1,"deletionMode":"Zone"}
...
{"level":"info","ts":1665001847.4786723,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","MissingProcesses"]}
{"level":"info","ts":1665001847.4787035,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001847.4787333,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001847.478752,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001847.4787745,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["SidecarUnreachable","MissingProcesses","PodPending"]}
...
{"level":"info","ts":1665001868.457372,"logger":"fdbclient","msg":"Fetch values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001872.4875183,"logger":"fdbclient","msg":"Done fetching values from FDB","namespace":"namespace","cluster":"foundationdb-cluster","key":"\ufffd\ufffd/status/json"}
{"level":"info","ts":1665001872.4885874,"logger":"controller","msg":"Check desired fault tolerance","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","expectedFaultTolerance":2,"maxZoneFailuresWithoutLosingData":2,"maxZoneFailuresWithoutLosingAvailability":2}
{"level":"info","ts":1665001872.4886274,"logger":"controller","msg":"Taking lock on cluster","namespace":"namespace","cluster":"foundationdb-cluster","action":"updating pods"}
{"level":"info","ts":1665001872.4886386,"logger":"controller","msg":"Deleting pods","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","zone":"26081535-vmss0000y5","count":1,"deletionMode":"Zone"}
...
{"level":"info","ts":1665001878.5913925,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","MissingProcesses","MissingPod"]}
{"level":"info","ts":1665001878.5913997,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","IncorrectConfigMap"]}
{"level":"info","ts":1665001878.5914106,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","IncorrectConfigMap"]}
{"level":"info","ts":1665001878.5914423,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","SidecarUnreachable","IncorrectConfigMap"]}
{"level":"info","ts":1665001878.591463,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["SidecarUnreachable","MissingProcesses","PodPending","IncorrectConfigMap"]}
...
{"level":"info","ts":1665001923.0192153,"logger":"controller","msg":"Check desired fault tolerance","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","expectedFaultTolerance":2,"maxZoneFailuresWithoutLosingData":1,"maxZoneFailuresWithoutLosingAvailability":1}
{"level":"info","ts":1665001923.0193546,"logger":"controller","msg":"Reconciliation terminated early","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.updatePods","requeueAfter":15,"message":"Reconciliation requires deleting pods, but deletion is currently not safe"}
...
{"level":"info","ts":1665001890.7436044,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","MissingProcesses","MissingPod"]}
{"level":"info","ts":1665001890.7436225,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","IncorrectConfigMap"]}
{"level":"info","ts":1665001890.7436473,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","IncorrectConfigMap"]}
{"level":"info","ts":1665001890.7436566,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec","SidecarUnreachable","IncorrectConfigMap"]}
{"level":"info","ts":1665001890.7436712,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["SidecarUnreachable","MissingProcesses","PodPending","IncorrectConfigMap"]}
...
{"level":"info","ts":1665001951.2949114,"logger":"controller","msg":"Check desired fault tolerance","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","expectedFaultTolerance":2,"maxZoneFailuresWithoutLosingData":0,"maxZoneFailuresWithoutLosingAvailability":0}
{"level":"info","ts":1665001951.2950678,"logger":"controller","msg":"Reconciliation terminated early","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.updatePods","requeueAfter":15,"message":"Reconciliation requires deleting pods, but deletion is currently not safe"}
...
{"level":"info","ts":1665001955.387233,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-2","state":"HasUnhealthyProcess","conditions":["MissingProcesses","PodPending"]}
{"level":"info","ts":1665001955.3872955,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-3","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001955.3873107,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-5","state":"HasUnhealthyProcess","conditions":["IncorrectPodSpec"]}
{"level":"info","ts":1665001955.3873217,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-6","state":"HasUnhealthyProcess","conditions":["SidecarUnreachable","MissingProcesses","PodPending"]}
{"level":"info","ts":1665001955.387333,"logger":"controller","msg":"Has unhealthy process group","method":"CheckReconciliation","namespace":"namespace","cluster":"foundationdb-cluster","processGroupID":"storage-9","state":"HasUnhealthyProcess","conditions":["SidecarUnreachable","MissingProcesses","PodPending","IncorrectConfigMap"]}
...
{"level":"info","ts":1665001983.75503,"logger":"controller","msg":"Check desired fault tolerance","namespace":"namespace","cluster":"foundationdb-cluster","reconciler":"updatePods","expectedFaultTolerance":2,"maxZoneFailuresWithoutLosingData":-1,"maxZoneFailuresWithoutLosingAvailability":-1}
{"level":"info","ts":1665001983.7551894,"logger":"controller","msg":"Reconciliation terminated early","namespace":"namespace","cluster":"foundationdb-cluster","subReconciler":"controllers.updatePods","requeueAfter":15,"message":"Reconciliation requires deleting pods, but deletion is currently not safe"}
```
</details>

**Comment by johscheuer** [comment]
> Hi! I reproduced this issue with version 1.8.1 of the fdb-kubernetes-operator, see the operator logs in the drop down.
> 
> **Setup**
> 
> ```
> Spec:
>   Database Configuration:
>     Logs:                             5
>     Proxies:                          3
>     redundancy_mode:                  triple
>     Storage:                          5
>     storage_engine:                   ssd
>   Minimum Uptime Seconds For Bounce:  600
>   Process Counts:
>     Log:        5
>     Stateless:  3
>     Storage:    5
> 
> Running Version:  6.3.12
> ```
> 
> To make sure that no recreated pods could be scheduled, I cordoned all the nodes. `kubectl cordon <node>`. I then triggered a rolling bounce pod update strategy (by changing the sidecar image tag).
> 
> We can see the reported fault tolerance from `status/json` is 2 just after two of the storage pods are detected as MissingProcesses (one of them also reporting PodPending). The kubernetes operator ends up deleting and recreating three pods (as pending), before the reported fault tolerance from `status/json` goes below 2.
> 
> End result is the cluster being unavailable, as a whole team of pods are not running (3 out of 5 storage pods are pending).
> 
> Operator Logs

Thanks for reporting I'll take a look and come back to that next week.

**Comment by manfontan** [comment]
🤔 not sure if this could be part of the problem. But I have realised that we are using FoundationDBCluster api version `v1beta1`. Maybe upgrading to `v1beta2` and enabling/configuring automatic replacement may help?

File: controllers/update_pods.go [hunk 3]
```diff
@@ -166,6 +175,22 @@ func shouldRequeueDueToTerminatingPod(pod *corev1.Pod, cluster *fdbv1beta2.Found
...
+// Because the internal.fault_tolerance relies in fdbcli status only we need to make sure it is
+// synced with the cluster status to prevent race conditions.
+func shouldRequeueDueToClusterStatusDrift(adminClient fdbadminclient.AdminClient, podMap map[string]*corev1.Pod) (bool, error) {
+	status, err := adminClient.GetStatus()
+	if err != nil {
+		return false, err
+	}
+	//Check that all processes in fdbcli status have a matching pod.
+	for _, p := range status.Cluster.Processes {
+		if _, ok := podMap[p.Locality[fdbv1beta2.FDBLocalityInstanceIDKey]]; !ok {
```
Comment (johscheuer) [inline controllers/update_pods.go:187]: I think this will probably not solve the actual issue, since a Pod that is in `Pending` state would be ignored here and Pods that are stuck in Pending (for whatever reason) would stay there and the operator would proceed deleting Pods once the data is redistributed. What we probably want/need is to implement a logic like `maxUnavailable ` from the Deployment (See: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#max-unavailable), that would ensure that the operator stops removing Pods after x Pods went into an unavailable state (in our case `MissingProcesses` or `PodPending`). I open an issue for this.

**Review by johscheuer — COMMENTED** [review]
Have you validated that the new code actually solves the case you've seen? Could you add some tests for the new logic?
